fix: rename incorrect keyword argument to `parse_query_string`

dodumosu opened the pull request:
`falconparser` uses `falcon.util.uri.parse_query_string`, but uses
an incorrect keyword argument name, `keep_blank_qs_values`. the correct
argument name is `keep_blank`

see: #448

dodumosu committed:
fix: rename incorrect keyword argument to `parse_query_string`

`falconparser` uses `falcon.util.uri.parse_query_string`, but uses
an incorrect keyword argument name, `keep_blank_qs_values`. the correct
argument name is `keep_blank`

see: #448

dodumosu committed:
feat: allow different keyword args for Falcon 1.x and 2.x

lafrech commented on src/webargs/falconparser.py:
```diff
@@ -13,6 +13,9 @@
 
 
+FALCON_VERSION = falcon.__version__[0]
+
```
We could copy compat.py and do

FALCON_VERSION_INFO = tuple(LooseVersion(falcon.__version__).version)

but I don't think there is much interest in doing so.

Let's keep it as is.

lafrech commented on src/webargs/falconparser.py:
```diff
@@ -13,6 +13,9 @@
 
+FALCON_VERSION = falcon.__version__[0]
+
+
```
On second thought, I'd write

    FALCON_VERSION_INFO = tuple(LooseVersion(falcon.__version__).version)

so that we get version as integer.

(We could also cast to `int`. I don't really mind.)

lafrech commented on src/webargs/falconparser.py:
```diff
@@ -65,9 +68,15 @@ def parse_form_body(req):
+            key = "keep_blank_qs_values"
+        elif FALCON_VERSION == "2":
+            key = "keep_blank"
+        kwargs = {key: req.options.keep_blank_qs_values}
```
I'd write

```py
        if FALCON_VERSION_INFO < 2:
            key = "keep_blank_qs_values"
        else:
            key = "keep_blank"
```

This way, it won't break in Falcon 3.

lafrech commented on src/webargs/falconparser.py:
```diff
@@ -65,9 +68,15 @@ def parse_form_body(req):
-                body, keep_blank_qs_values=req.options.keep_blank_qs_values
+                body, **kwargs
             )
     return {}
```
A single line should be enough.

lafrech reviewed (changes requested):
Good.

I think it makes sense to maintain compatibility with Falcon 1. Flacon 2 is really recent and backward compat is cheap.

Please also add yourself to AUTHORS.

dodumosu committed:
chore: add changes requested by project lead

this commit mostly consists of stylistic changes, for example
using `distutils.version.LooseVersion` instead of the raw Falcon
version string.

other than that, the major change is dealing all Falcon versions
greater than 1.X as using the updated keyword arg, and Falcon 1.x
versions as using the legacy keyword arg (as opposed to just
Falcon 1.x and 2.x).

dodumosu committed:
chore: add name to AUTHORS

lafrech commented on src/webargs/falconparser.py:
```diff
@@ -13,7 +15,7 @@
 
 
-FALCON_VERSION = falcon.__version__[0]
+FALCON_VERSION = tuple(LooseVersion(falcon.__version__).version)
```
s/FALCON_VERSION/FALCON_VERSION_INFO/

Or at least be consistent with below.

lafrech reviewed (changes requested):
Great. Just one typo to fix.

dodumosu committed:
fix: fix typo

lafrech commented on src/webargs/falconparser.py:
```diff
@@ -65,10 +70,14 @@ def parse_form_body(req):
+            key = "keep_blank_qs_values"
+        else:
+            key = "keep_blank"
+        kwargs = {key: req.options.keep_blank_qs_values}
```
The whole `if FALCON_VERSION_INFO[0] < 2:` block could be moved under the `if body` condition.

```py
        if body:
            if FALCON_VERSION_INFO[0] < 2:
                key = "keep_blank_qs_values"
            else:
                key = "keep_blank"
            return parse_query_string(body, **{key: req.options.keep_blank_qs_values})
```

lafrech reviewed (changes requested):
One last change I didn't think of before.

Thanks.

dodumosu committed:
chore: refactor block

lafrech commented:
If we mean to support both Flacon 1 and 2, we should ensure the tests test both version. @dodumosu if you're not comfortable with modifying CI tests config, we'll do it when we get the time.

Is Flacon 1 still maintained? Shall we drop support for it in the next webargs major version ?

Nateyo commented:
As far as I know, most of the development going on with falcon is now 2.0+. This PR should likely update the setup.py as well since right now it specifies `falcon>=1.4.0,<2.0`. There's actually another problem with args to `req.stream.read()` which I can submit a PR for, building on top of @dodumosu's changes

lafrech commented:
What about keeping webargs 5.x Falcon 1.x and dropping Falcon 1.x in webargs 6.0?

|webargs|Falcon|
|-|-|
|5|1 only|
|6|2 only|

lafrech commented:
@sloria, we could

- [x] merge #420
- [x] drop Python 2
- [ ] apply the change in this PR but just keep Flacon 2 compatibility, update setup.py (and CI config if needed)
- [ ] fix `req.stream.read()` as per [this comment](https://github.com/marshmallow-code/webargs/pull/449#issuecomment-569941269)

We could publish beta releases along the way.

sloria commented:
@lafrech I'm good with that plan

lafrech commented:
Guys, I just merged #420.

You may rebase this and drop Falcon 1 (or start from scratch if it is easier), and send another PR to fix `req.stream.read()`.

Thanks.

lafrech commented:
~Better wait until #454 is merged to avoid merge conflicts.~ Done.

lafrech commented:
@dodumosu @Nateyo are you willing to tackle this?

No pressure, just asking. I could rebase the change in this PR when I get the time, but I don't know what @Nateyo has in mind regarding `req.stream.read()` and I don't use Falcon so ideally I'd rather defer that to you.

Basically, change setup.py to require falcon>=2 and do all required changes (the two expressed above + others if needed).

Nateyo commented:
@lafrech Sure, I can work on it this week. I can cherry pick these commits and submit a whole new PR to the dev branch? Would that be easier since you want 5.x to keep Falcon 1 compat?

lafrech commented:
Great. Thanks @Nateyo.

Further dev on webargs 5 will take place in branch `5.x-line`. We're dropping Falcon 1 in `dev` branch. I wouldn't cherry-pick and just start from scratch as the changeset is much smaller when dropping Falcon 2 than when keeping compatibility. You may submit a single PR with the whole Falcon 2 move.

lafrech commented:
Superseded by #459.
